compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -1481,7 +1497,12 @@ private boolean validateRequiredAndReadonlyFields(BRecordType type,
                         types.isAssignable(typePosPair.type, symTable.readonlyType)) {
                     continue;
                 }
-                dlog.error(typePosPair.pos, DiagnosticCode.INVALID_FIELD_FOR_READONLY_RECORD_FIELD, fieldName);
+
+                if (mutableRecordType && !Symbols.isFlagOn(field.type.flags, Flags.READONLY)) {
```

Comment:
The `!Symbols.isFlagOn(field.type.flags, Flags.READONLY)` check is not relevant here right? IIUC, this checks if there's a `readonly` field in a mutable record type? Plus, isn't it the same check as in the if block above? Is the field being `readonly` and the field's type being `readonly` two different things?

Reply:
Yeah they're different.

So there are three possible scenarios.

1. The expected type is `T & readonly` where T is the record type. Meaning, the value itself will be immutable and whatever fields that are specified should also be immutable. This is what is checked with `mutableRecordType`.

```ballerina
Foo & readonly f = {};
```

2. A field's type is `T & readonly`. This means that whatever value set for that field will have to be immutable. That's the second check.

```ballerina
type Foo record {|
    Bar & readonly b;
|};
``` 
And you can also update the `b` field in a `Foo` value after creating  the record value, but the `Bar` value must always be immutable.

3. Then there's `readonly` fields where in addition to 2) there's also the constraint that the value cannot be updated once it is set.

```ballerina
type Foo record {|
    readonly Bar b;
|};
```

We only want to log an error here for 3.